Canary upgrade docs (#512)
banzaicloud/istio-operator: docs/upgrade/canary-upgrade.md

# Istio canary upgrade guide

## High level flow

Let us suppose that we have an Istio 1.6 control plane running in our cluster and we would like to upgrade to an Istio 1.7 control plane.

> Here are the high level steps necessary to accomplish this through the canary upgrade flow:

Reviewer: ```suggestion
To make sure that we are limiting the potential blast radius of the upgrade, we introduced the Canary upgrade flow. This consists of the following steps:
```

1. Deploy an Istio 1.7 control plane *next* to the Istio 1.6 control plane
1. Migrate the data plane *gradually* to the new control plane
1. When migration is fully finished, uninstall the Istio 1.6 control plane

![Istio control plane canary upgrade](istio-cp-canary-upgrade.gif)

This canary upgrade model is much safer than the in-place upgrade workflow, mainly because it can be done gradually, at your own pace.

> If issues arise they will probably only affect a small portion of your applications, initially, and, if that does happen, it should be much easier to rollback only a part of your application for use with the original control plane.

Reviewer: ```suggestion
If issues arise they will only affect a small portion of your applications, that you have selected as the initial scope for the upgrade. If that does happen, it will be much easier to rollback only a part of your application for use with the original control plane.
```

So our starting point is a running Istio operator and Istio 1.6 control plane, with applications using that control plane version.

1. Deploy another Istio operator (alongside the previous one), which has Istio 1.7 support
1. Apply a new Istio CR with Istio version 1.7 and let the operator reconcile an Istio 1.7 control plane

   It is recommended that you turn off gateways for the new control plane and migrate the 1.6 gateway pods to the 1.7 control plane. That way the ingress gateway's typically LoadBalancer-type service will not be recreated, neither will the cloud platform-managed LoadBalancer, and hence the IP address won't change either.

   > This new control plane is often referred to as a *canary*, but, in practice, it is advisable that this be named based on Istio version, as it will remain on the cluster for the long term.

1. Migrate gateways to use the new Istio 1.7 control plane

   This has been intentionally left as a manual step during the new control plane installation with the Istio operator, so that users can make sure that they're ready to use the control plane for their gateways, and only migrate after.

1. Migrate data plane applications

>    It is recommended that you perform this migration gradually.

Reviewer: ```suggestion
   It is recommended that for safety reasons you perform this migration gradually.
```

>    It's important to keep in mind during the data plane migration that the two Istio control planes share trust (because they use the same certificates for encrypted communication). Therefore, when an application pod that still uses Istio's 1.6 control plane calls another pod that already uses Istio's 1.7 control plane, the encrypted communication will succeed because of that shared trust.

Reviewer: ```suggestion
  During the data plane migration, it is important to keep in mind that the two Istio control planes share trust (because they use the same certificates for encrypted communication). Therefore, when an application pod that still uses Istio's 1.6 control plane calls another pod that already uses Istio's 1.7 control plane, the encrypted communication will succeed because of that shared trust.
```

>    That's why this migration can be performed on a namespace by namespace basis and the communication of the pods won't be affected.

Reviewer: ```suggestion
   That's why this migration can be performed on a namespace by namespace basis and the **communication** of the pods **won't be affected.**
```

1. Delete the Istio 1.6 control plane

>    Once the migration is finished, and we've made sure that our applications are working properly in conjunction with the new control plane, the older Istio 1.6 control plane can be safely deleted.

Reviewer: ```suggestion
   Once the migration is finished, and you've made sure that your applications are working properly in conjunction with the new control plane, the older Istio 1.6 control plane can be safely deleted.
```

Reviewer: You were using you before.

>    It's recommended that you take some time to make sure that everything is working on the new control plane before turning off the old one; it's only an `istiod` deployment running with practically no data plane pods attached.

Reviewer: ```suggestion
   It's recommended that you take some time to make sure that everything is working on the new control plane before turning off the old one. The overhead of doing so is minimal as it's only an `istiod` deployment running with practically no data plane pods attached.
```

Reviewer: I would remove this as it has no real meaning, and I feel like you don't want to go into details here ;-):

practically no data plane pods attached.

   ```

1. Apply an `Istio` Custom Resource and let the operator reconcile the Istio 1.6 control plane.

   ```bash
   $ kubectl apply -n istio-system -f https://raw.githubusercontent.com/banzaicloud/istio-operator/release-1.6/config/samples/istio_v1beta1_istio.yaml
   ```

1. Check that the Istio 1.6 control plane is deployed.

   ```bash
   $ kubectl get po -n=istio-system
   NAME                                    READY   STATUS    RESTARTS   AGE
   istio-ingressgateway-55b89d99d7-4m884   1/1     Running   0          17s
   istio-operator-v16x-0                   2/2     Running   0          57s
   istiod-5865cb6547-zp5zh                 1/1     Running   0          29s
   ```

### Deploy demo app

1. Create two namespaces for the demo apps.

   ```bash
   $ kubectl create ns demo-a
   $ kubectl create ns demo-b
   ```

1. Add those namespaces to the mesh.

   ```bash
   $ kubectl patch istio -n istio-system istio-sample --type=json -p='[{"op": "replace", "path": "/spec/autoInjectionNamespaces", "value": ["demo-a", "demo-b"]}]'
   ```

> 1. Make sure that the namespaces are labeled for sidecar injection (if not, wait a few seconds).

Reviewer: ```suggestion
1. Make sure that the namespaces are labeled for sidecar injection (if not, wait a few seconds, then please re-check the namespaces).
```

   spec:
     ports:
     - name: http
       port: 80
       targetPort: 8080
     selector:
       k8s-app: app-a
   ```

   ```yaml
   apiVersion: apps/v1
   kind: Deployment
   metadata:
     name: app-b
     labels:
       k8s-app: app-b
     namespace: demo-b
   spec:
     replicas: 1
     selector:
       matchLabels:
         k8s-app: app-b
     template:
       metadata:
         labels:
           k8s-app: app-b
       spec:
         terminationGracePeriodSeconds: 2
         containers:
         - name: echo-service
           image: k8s.gcr.io/echoserver:1.10
           ports:
           - containerPort: 8080
   ---
   apiVersion: v1
   kind: Service
   metadata:
     name: app-b
     labels:
       k8s-app: app-b
     namespace: demo-b
   spec:
     ports:
     - name: http
       port: 80
       targetPort: 8080
     selector:
       k8s-app: app-b
   ```

### Test communication

1. Save the application pod names for easier access.

   ```bash
   $ APP_A_POD_NAME=$(kubectl get pods -n demo-a -l k8s-app=app-a -o=jsonpath='{.items[0].metadata.name}')
   $ APP_B_POD_NAME=$(kubectl get pods -n demo-b -l k8s-app=app-b -o=jsonpath='{.items[0].metadata.name}')
   ```

> 1. Test to make sure that app-a can access app-b.

Reviewer: ```suggestion
1. Test if app-a can access app-b.
```

   ```bash
   $ kubectl exec -n=demo-a -ti -c echo-service $APP_A_POD_NAME -- curl -Ls -o /dev/null -w "%{http_code}" app-b.demo-b.svc.cluster.local
   200
   ```

> 1. Test to make sure that app-b can access app-a.

Reviewer: ```suggestion
1. Test if app-b can access app-a.
```

1. Deploy an Istio operator version, which can install an Istio 1.7 control plane.

   ```bash
   $ helm install istio-operator-v17x --create-namespace --namespace=istio-system --set-string operator.image.tag=0.7.1 banzaicloud-stable/istio-operator
   ```

1. Apply an `Istio` Custom Resource and let the operator reconcile the Istio 1.7 control plane.

   ```bash
   $ kubectl apply -n istio-system -f https://raw.githubusercontent.com/banzaicloud/istio-operator/release-1.7/config/samples/istio_v1beta1_istio.yaml
   ```

1. Check that the Istio 1.7 control plane is also deployed.

   ```bash
   $ kubectl get po -n=istio-system
   NAME                                        READY   STATUS    RESTARTS   AGE
   istio-ingressgateway-55b89d99d7-4m884       1/1     Running   0          6m38s
   istio-operator-v16x-0                       2/2     Running   0          7m18s
   istio-operator-v17x-0                       2/2     Running   0          76s
   istiod-676fc6d449-9jwfj                     1/1     Running   0          10s
   istiod-istio-sample-v17x-7dbdf4f9fc-bfxhl   1/1     Running   0          18s
   ```

### Migrate data plane

#### Migrate ingress

> 1. Move the ingress gateway so as to utilize the new Istio 1.7 control plane.

Reviewer: ```suggestion
1. Change the ingress gateway so that it utilizes the new Istio 1.7 control plane.
```

   ```bash
   $ kubectl patch mgw -n istio-system istio-ingressgateway --type=json -p='[{"op": "replace", "path": "/spec/istioControlPlane/name", "value": "istio-sample-v17x"}]'
   ```

#### Migrate first namespace

> 1. Label the first namespace for the new control plane.

Reviewer: ```suggestion
1. Label the first namespace so that all workloads there utilize the new control plane.
```

   ```bash
   $ kubectl label ns demo-a istio-injection- istio.io/rev=istio-sample-v17x.istio-system
   ```

>    The new `istio.io/rev` label needs to be used for the new revisioned control planes to be able to perform sidecar injection.

Reviewer: ```suggestion
   The new `istio.io/rev` label needs to be used for the new revisioned control planes to indicate that it should perform sidecar injection.
```

   The `istio-injection` label must be removed because it takes precedence over the `istio.io/rev` label for backward compatibility.

1. Make sure that the labeling is correct.

   ```bash
   $ kubectl get ns demo-a -L istio-injection -L istio.io/rev
   NAME     STATUS   AGE   ISTIO-INJECTION   REV
   demo-a   Active   12m                     istio-sample-v17x.istio-system
   ```

1. Restart the pod in the namespace.

   ```bash
   $ kubectl rollout restart deployment -n demo-a
   ```

> 1. Make sure that now the new 1.7 sidecar proxy is used for the new pod.

Reviewer: ```suggestion
1. Make sure that the new 1.7 sidecar proxy is used for the new pod.
```

   ```bash
   $ APP_A_POD_NAME=$(kubectl get pods -n demo-a -l k8s-app=app-a -o=jsonpath='{.items[0].metadata.name}')
   $ kubectl get po -n=demo-a $APP_A_POD_NAME -o yaml | grep istio/proxyv2:
       image: docker.io/istio/proxyv2:1.7.0
       image: docker.io/istio/proxyv2:1.7.0
       image: docker.io/istio/proxyv2:1.7.0
       image: docker.io/istio/proxyv2:1.7.0
   ```

#### Test communication

Let's make sure that encrypted communication still works between pods that use different control planes.

> This is a very important part and the reason why the data plane migration can be done gradually, because the communication works even between pods on different control planes.

Reviewer: ```suggestion
The reason why the data plane migration can be done gradually is that the communication works even between pods on different control planes. In this step please it will be verified that the communication works in such a way, making the upgrade flow a safe one.
```

> Remember that the pod in the `demo-a` namespace are already on Istio 1.7 control plane, but the pod in `demo-b` is still on Istio 1.6.

Reviewer: ```suggestion
Remember that the pod(s) in the `demo-a` namespace are already on the Istio 1.7 control plane, but the pod in `demo-b` is still using the Istio 1.6 version.
```

1. Save the application pod names for easier access.

   ```bash
   $ APP_A_POD_NAME=$(kubectl get pods -n demo-a -l k8s-app=app-a -o=jsonpath='{.items[0].metadata.name}')
   $ APP_B_POD_NAME=$(kubectl get pods -n demo-b -l k8s-app=app-b -o=jsonpath='{.items[0].metadata.name}')
   ```

> 1. Test to make sure that app-a can access app-b.

Reviewer: ```suggestion
1. Test if app-a can access app-b.
```

   ```bash
   $ kubectl exec -n=demo-a -ti -c echo-service $APP_A_POD_NAME -- curl -Ls -o /dev/null -w "%{http_code}" app-b.demo-b.svc.cluster.local
   200
   ```

> 1. Test to make sure that app-b can access app-a.

Reviewer: ```suggestion
1. Test if app-b can access app-a.
```

   ```bash
   $ kubectl exec -n=demo-b -ti -c echo-service $APP_B_POD_NAME -- curl -Ls -o /dev/null -w "%{http_code}" app-a.demo-a.svc.cluster.local
   200
   ```

#### Migrate second namespace

> 1. Label the second namespace for the new control plane.

Reviewer: Please see the recommendations above.

   ```bash
   $ kubectl label ns demo-b istio-injection- istio.io/rev=istio-sample-v17x.istio-system
   ```

1. Make sure that the labeling is correct.

   ```bash
   $ kubectl get ns demo-b -L istio-injection -L istio.io/rev
   NAME     STATUS   AGE   ISTIO-INJECTION   REV
   demo-b   Active   19m                     istio-sample-v17x.istio-system
   ```

1. Restart the pod in the namespace.

   ```bash
   $ kubectl rollout restart deployment -n demo-b
   ```

1. Make sure that now the new 1.7 sidecar proxy is used for the new pod.

   ```bash
   $ APP_B_POD_NAME=$(kubectl get pods -n demo-b -l k8s-app=app-b -o=jsonpath='{.items[0].metadata.name}')
   $ kubectl get po -n=demo-b $APP_B_POD_NAME -o yaml | grep istio/proxyv2:
       image: docker.io/istio/proxyv2:1.7.0
       image: docker.io/istio/proxyv2:1.7.0
       image: docker.io/istio/proxyv2:1.7.0
       image: docker.io/istio/proxyv2:1.7.0
   ```

### Uninstall the Istio 1.6 control plane

> When the data plane is fully migrated and we made sure it works as expected, we can delete the "old" Istio 1.6 control plane now.

Reviewer: ```suggestion
When the data plane is fully migrated to the 1.7 version and you made sure that it works as expected, we can delete the "old" Istio 1.6 control plane.
```
